### ISSUE TITLE
Extend the type annotation of `catch` argument of `study.optimize`

### ISSUE THREAD

**Issue by nzw0301** [optuna/optuna#3230]
### Motivation

As reported by https://github.com/optuna/optuna/issues/2163, the current type of `catch` argument of `study.optimize`, tuple, might be inconvenient for users.

### Description

I'd like to suggest changing the argument's type as follows

- single `Exception`: `catch=RuntimeWarning `
- sequence of `Exception`: `catch=[RuntimeWarning]` or `catch=(RuntimeWarning, InvalidArgumentError)`

So the type annotation might be ~`Optional[Sequence[Type[Exception]], Type[Exception]]]`~, `Optional[Union[Sequence[Type[Exception]], Type[Exception]]]` (thanks @BasLaa).

### Alternatives (optional)

The `states` argument of [`study.get_trials`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.get_trials) might have the same issue because it requires a tuple of trial states.

I'd like to hear Optuna community opinions.

**Comment by nzw0301** [comment]
Indeed, some examples use a list rather than tuple for `states` of `study.get_trials`. https://github.com/optuna/optuna-examples/blob/07c77d8baa17e18ee0b85435df080c11749c6777/pytorch/pytorch_simple.py#L128-L129

**Comment by BasLaa** [comment]
Wouldn't we do something like 
`Optional[Union[Sequence[Type[Exception]], Type[Exception]]]`?
or equivalently `Optional[Sequence[Type[Exception]] | Type[Exception]]`
This would indicate either None, a sequence or simply an exception.
The default value sensibly would be None I believe.

I can change it quickly?

**Comment by BasLaa** [comment]
Similarly for states: `Optional[Sequence[optuna.trial._state.TrialState] | optuna.trial._state.TrialState]`

**Comment by himkt** [comment]
> Wouldn't we do something like Optional[Union[Sequence[Type[Exception]], Type[Exception]]]?

I think so too (`|` operator is available from Python 3.10 so we cannot use it yet).

> Similarly for states: Optional[Sequence[optuna.trial._state.TrialState] | optuna.trial._state.TrialState]

I'd like to be careful for adding the same change to `states` because `states` is a plural.
It looks enough to me that changes from `Optional[Tuple[TrialState, ...]]` to `Optional[Sequence[TrialState]]`.

**Comment by BasLaa** [comment]
Ah right, good that you mention that `|` is from 3.10.
Right, that makes sense for states, I'll do that then.
I will work on implementing this and link you the PR.

**Comment by BasLaa** [comment]
This will require all the classes that build on the Study class to have the type annotation changed too.
These are:
`optimize.py` for the lightgbm tuner
`_hyperband.py` in pruners
`_base.py` in storages (get_all_trials method)

**Comment by nzw0301** [comment]
I've read the codebase a little bit to investigate this issue. Let me summarise my current thoughts as follows:

### `catch`

This argument is used in https://github.com/optuna/optuna/blob/ac4ea8d0c74726f8a603ba2cb0bfb7f4112f736e/optuna/study/_optimize.py#L263

So we need to change the type of `catch` to `tuple` before evaluating the line.

In addition, I originally proposed adding `Optinal` because it was supposed to be consistent with `states` argument of `study.get_trails`. However, `None` of `states` means all states, so it might be wrong to make them consistent, sorry for that.

Now, I refer to using `Union[Sequence[Type[Exception]], Type[Exception]` or `Sequence[Type[Exception]]`.


### `states`

This argument is used to filter the trials by their state attribute. For example, 

https://github.com/optuna/optuna/blob/6e54a7468f52b0f89d090b49c87b48f1c976608f/optuna/storages/_cached_storage.py#L353

As raised in https://github.com/optuna/optuna/issues/3230#issuecomment-1027481524, accepting a single `TrialState` might make the codebase a little bit complicated. So we can relax this type annotation to `Optional[Container[optuna.trial._state.TrialState]]`, where `Container` supports `in` operation.

**Comment by himkt** [comment]
@nzw0301

> I prefer relaxing this type annotation to Optional[Container[optuna.trial._state.TrialState]]

Do you think it's better to allow other container objects such as `set`? It feels little bit too general to me.
Isn't it enough to allow `list` (and related datatypes) by using `Sequence`?

**Comment by nzw0301** [comment]
Good point. 

`Container` sounds more accurate than `Sequence` in the current implementation at least for me. In addition, `set` is a commonly used type, which is not the instance of `Sequence` as you mentioned. 

Note that https://github.com/optuna/optuna/blob/4d0bbcc49ebc418b6dc410dee7c1edc0e6c1492f/optuna/integration/cma.py#L479 uses set for the same purpose.

**Comment by himkt** [comment]
I see. Thank you for explaining. 📝

**Comment by himkt** [comment]
@BasLaa, thank you for improving `states`!

We wait another PR to modify `catch` as well (ref. https://github.com/optuna/optuna/pull/3359#issuecomment-1060454209). Please feel free to mention if I can help you!

**Comment by BasLaa** [comment]
Yep, it's coming! It might be a day or two, uni is busy!

**Comment by BasLaa** [comment]
Apologies for this delay, it's my exam period in university so everything is taking a while. At the latest it'll be Saturday.

**Comment by himkt** [comment]
Hi @BasLaa, please tell me if I can help you.

**Comment by HideakiImamura** [comment]
This issue is still an open. Please feel free to submit a PR.

Let me remove the good-first-issue label since this issue has a slightly larger context.

**Comment by gen740** [comment]
As a user, I want to write code like this.
```
study.optimize(objective, catch=SomeException, n_trials=100)
```

But currently, we should
```
study.optimize(objective, catch=(SomeException,), n_trials=100)
```

So adding support of single Exception catch would be huge improvement. ( `Union[Tuple[Type[Exception], ...], Type[Exception]]`)

**Comment by hvy** [comment]
Should we keep this issue open, even though #4098 has been merged?

**Comment by nzw0301** [comment]
Thanks! I think we can close this issue thanks to #4098.
